File: src/test/java/com/pubnub/api/PNConfigurationTest.java
package com.pubnub.api;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;

class PNConfigurationTest {

    @Test
    void should_set_uuid_creating_PNConfiguration() throws PubNubException {
        String userId01value = "userId01";
        PNConfiguration pnConfiguration = new PNConfiguration(new UserId(userId01value));

        assertEquals(userId01value, pnConfiguration.getUuid());
    }

    @Test
    void can_setUserId() throws PubNubException {
        PNConfiguration pnConfiguration = new PNConfiguration(new UserId("userId01"));
        UserId newUserId = new UserId("newUserId");
        pnConfiguration.setUserId(newUserId);

        assertEquals(newUserId.getValue(), pnConfiguration.getUserId().getValue());
    }

    @Test
    void can_getUserId() throws PubNubException {
        String userId01value = "userId01";
        PNConfiguration pnConfiguration = new PNConfiguration(new UserId(userId01value));
        UserId retrievedUserId = pnConfiguration.getUserId();

        assertEquals(userId01value, retrievedUserId.getValue());
    }

    @Test
    void can_reset_userId_to_non_empty_string() throws PubNubException {
        PNConfiguration pnConfiguration = new PNConfiguration(new UserId("userId01"));
        String newUserIdValue = "newUserId";
        pnConfiguration.setUserId(new UserId(newUserIdValue));
        assertEquals(newUserIdValue, pnConfiguration.getUserId().getValue());
    }

    @Test
    void should_throw_exception_when_uuid_is_empty_string() {
        Assertions.assertThrows(PubNubException.class, () -> new PNConfiguration(""));
    }

    @Test
    void should_throw_exception_when_uuid_is_null_string() {
        Assertions.assertThrows(PubNubException.class, () -> {
            new PNConfiguration((String) null);
        });
    }
}

File: kotlin/src/test/kotlin/com/pubnub/api/PNConfigurationTest.kt
package com.pubnub.api

import org.junit.Assert.assertEquals
import org.junit.Test

class PNConfigurationTest {
    @Suppress("DEPRECATION")
    @Test
    fun pnConfigurationGeneratesPnsdkWithSuffixes() {
        val name1 = "key1"
        val suffix1 = "value1/1.0.0"
        val name2 = "key2"
        val suffix2 = "value2/2.0.0"
        val suffix11 = "value3/2.0.0"

        val pnConfiguration = object :
            PNConfiguration(userId = UserId(PubNub.generateUUID())) {
            init {
                addPnsdkSuffix(name1 to suffix1, name2 to suffix2)
                addPnsdkSuffix(mapOf(name1 to suffix11))
            }
        }
        val version = "someVersion"
        val generatedPnsdk = pnConfiguration.generatePnsdk(version)
        assertEquals("PubNub-Kotlin/$version $suffix11 $suffix2", generatedPnsdk)
    }

    @Test(expected = PubNubException::class)
    fun setUserIdToEmptyString() {
        PNConfiguration(userId = UserId(""))
    }

    @Test(expected = PubNubException::class)
    fun resetUserIdToEmptyString() {
        val config = PNConfiguration(userId = UserId(PubNub.generateUUID()))
        config.userId = UserId("")
    }

    @Test
    fun resetUserIdToNonEmptyString() {
        val config = PNConfiguration(userId = UserId(PubNub.generateUUID()))
        val newUserId = UserId(PubNub.generateUUID())
        config.userId = newUserId

        assertEquals(newUserId, config.userId)
    }

    @Suppress("DEPRECATION")
    @Test(expected = PubNubException::class)
    fun setUUIDToEmptyString() {
        PNConfiguration("")
    }

    @Suppress("DEPRECATION")
    @Test(expected = PubNubException::class)
    fun resetUUIDToEmptyString() {
        val config = PNConfiguration(PubNub.generateUUID())
        config.uuid = ""
    }

    @Suppress("DEPRECATION")
    @Test
    fun resetUUIDToNonEmptyString() {
        val config = PNConfiguration(PubNub.generateUUID())
        val newUUID = PubNub.generateUUID()
        config.uuid = newUUID

        assertEquals(newUUID, config.userId.value)
    }
}
